plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -0,0 +1,129 @@
+    private long currentOffset;
+
+    S3RetryingInputStream(S3BlobStore blobStore, String blobKey) throws IOException {
+        this.blobStore = blobStore;
+        this.blobKey = blobKey;
+        this.maxAttempts = blobStore.getMaxRetries() + 1;
+        currentStream = openStream();
+    }
+
+    private InputStream openStream() throws IOException {
+        try (AmazonS3Reference clientReference = blobStore.clientReference()) {
+            final GetObjectRequest getObjectRequest = new GetObjectRequest(blobStore.bucket(), blobKey);
+            if (currentOffset > 0) {
+                getObjectRequest.setRange(currentOffset);
+            }
+            final S3Object s3Object = SocketAccess.doPrivileged(() -> clientReference.client().getObject(getObjectRequest));
+            return s3Object.getObjectContent();
+        } catch (final AmazonClientException e) {
+            if (e instanceof AmazonS3Exception) {
+                if (404 == ((AmazonS3Exception) e).getStatusCode()) {
+                    throw new NoSuchFileException("Blob object [" + blobKey + "] not found: " + e.getMessage());
+                }
+            }
+            throw e;
+        }
+    }
+
+    @Override
+    public int read() throws IOException {
+        int attempt = 0;
+        while (true) {
+            attempt += 1;
+            try {
+                final int result = currentStream.read();
+                currentOffset += 1;
+                return result;
+            } catch (IOException e) {
+                if (attempt >= maxAttempts) {
+                    throw e;
+                }
+                logger.debug(new ParameterizedMessage("failed reading [{}/{}] at offset [{}], attempt [{}] of [{}], retrying",
+                    blobStore.bucket(), blobKey, currentOffset, attempt, maxAttempts), e);
+                currentStream.close();
+                currentStream = openStream();
+            }
+        }
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        int attempt = 0;
+        while (true) {
+            attempt += 1;
+            try {
+                final int bytesRead = currentStream.read(b, off, len);
+                if (bytesRead == -1) {
+                    return -1;
+                }
+                currentOffset += bytesRead;
+                return bytesRead;
+            } catch (IOException e) {
+                if (attempt >= maxAttempts) {
+                    throw e;
+                }
+                logger.debug(new ParameterizedMessage("failed reading [{}/{}] at offset [{}], attempt [{}] of [{}], retrying",
+                    blobStore.bucket(), blobKey, currentOffset, attempt, maxAttempts), e);
+                currentStream.close();
+                currentStream = openStream();
+            }
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        currentStream.close();
+        super.close();
```

Comment:
NIT: seems unnecessary to call noop `InputStream#close`?

Reply:
Thanks ad43a77c519e60be1b12d3e1158d515b93813a6a.